source/extensions/http/cache/file_system_http_cache/DESIGN.md

```diff
@@ -0,0 +1,37 @@
+[ ] Cache should expose counters for timing information (eviction thread idle, eviction thread busy)
+[ ] Cache should expose gauges for total size stored.
+[ ] Cache should optionally expose histograms for insert and lookup latencies.
+[x] Cache should index by the request route *and* a key generated from headers that may affect the outcome of a request (See [allowed_vary_headers](https://www.envoyproxy.io/docs/envoy/latest/api-v3/extensions/filters/http/cache/v3/cache.proto.html))
+[ ] Cache should allow a tree structure of folders, to avoid filesystem performance issues with overcrowded directories.
+
+## Storage design
+
+* The only state stored in memory is that a cache entry is in the process of being written; this allows other requests for the same resource to avoid creating duplicate write operations.
+* The cache can be configured with a maximum number of cache entry files, thereby effectively enforcing a maximum number of files per path.
+* A new cache entry that causes the cache to exceed the configured maximum size or maximum number of entries triggers the eviction thread to evict sufficient LRU entries to bring it back below the threshold\[s\] exceeded.
+* Each cache entry file starts with [a fixed structure header followed by a serialized proto](cache_file_header.proto), followed by proto-serialized headers, raw body and proto-serialized trailers.
+* Cache entry files are named `cache-` followed by a stable hash key for the entry.
+* (When implemented) the tree structure of folders is simply one level deep of folders named `cache-0000`, `cache-0001` etc. as four-digit hexadecimal numbers up to the configured number of subdirectories. Cache files are placed in a folder according to a short stable hash of their key. On cache startup, any cache entries found to be in the wrong folder (as would be the case if the number of folders was reconfigured) will simply be removed.
+
+## Discussions
+
+<a name="thundering-herd"></a>
+### Thundering herd
+
+The current implementation, if there are multiple requests for the same resource before the cache is populated, has only one of them perform an insert operation to the cache, and the rest simply bypass the cache. This can cause the "thundering herd" problem - if requests come in bursts the cache will not protect the upstream from that load.
+
+One possible solution would be to have all requesters for the same cache entry stream as the cache entry is written. However, if we do that, and the inserting stream gets closed prematurely, all the dependent streams would be forced to drop their also-incomplete responses.
+
+Another possible solution is to have secondary requesters wait until the cache entry is populated or abandoned before deciding whether to become cache readers or inserters (or bypass, if it turns out to be uncacheable). The downside of this option is that for large content, the dependent clients won't start streaming at all until the first client *finishes* streaming.
```

Comment:
that makes sense, and wasn't something we had considered an issue as video was not a target.

Maybe the initial requester can put the lock object into one of several states:
 * in-progress, size unknown
 * headers-complete; content-length known and less than a threshold
 * headers-complete, content-length known and more than a threshold
 * headers-complete, chunked encoding

In this case we can block secondary requests until we get the headers, and then we can decide: chunked or large content-length means "secondaries don't wait".

Reply:
Added paraphrase of this to the doc. Seems like a good idea in the short term, as any of the streaming-capable options seem like a better idea but involve significantly more effort up-front.